File: src/main/java/com/morak/performancetracker/context/TestMetadata.java
package com.morak.performancetracker.context;

import java.util.Objects;

public class TestMetadata {

    public static final TestMetadata ROOT = new TestMetadata("ROOT", "ROOT");

    private String className;
    private String methodName;

    public TestMetadata(String className, String methodName) {
        this.className = className;
        this.methodName = methodName;
    }

    public String getClassName() {
        return className;
    }

    public String getMethodName() {
        return methodName;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        TestMetadata that = (TestMetadata) o;
        return Objects.equals(className, that.className) && Objects.equals(methodName, that.methodName);
    }

    @Override
    public int hashCode() {
        return Objects.hash(className, methodName);
    }

    @Override
    public String toString() {
        return "TestMetadata{" +
                "className='" + className + '\'' +
                ", methodName='" + methodName + '\'' +
                '}';
    }
}

File: src/main/kotlin/com/morak/performancetracker/context/TestMetadata.kt
package com.morak.performancetracker.context

import com.morak.performancetracker.utils.substringFrom
import com.morak.performancetracker.utils.substringUntil

class TestMetadata(private val name: String) {
    val parent: TestMetadata
        get() = TestMetadata(name.substringUntil('@'))

    val self: String
        get() = name.substringFrom('@')

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as TestMetadata

        return name == other.name
    }

    override fun hashCode(): Int {
        return name.hashCode()
    }

    override fun toString(): String {
        return "TestMetadata{" +
                "name='" + self + '\'' +
                '}'
    }


    companion object {
        private const val ROOT_NAME = "ROOT"
        private const val NAME_DELIMITER = "@"

        val ROOT = TestMetadata(ROOT_NAME)

        fun of(vararg names: String): TestMetadata {
            val name = ROOT_NAME + NAME_DELIMITER + names.reduce { acc, cur -> acc + NAME_DELIMITER + cur }
            return TestMetadata(name)
        }
    }
}
